compiler/rustc_data_structures/src/sync.rs

```diff
@@ -299,48 +340,125 @@ cfg_if! {
+                parallel!([$dyn_send], impl $fblock [$block, $($c,)*] [$($rest),*])
+            };
+            ([$dyn_send: expr], impl $fblock:tt [$($blocks:tt,)*] []) => {
+                {
+                    let dyn_send = IsSend::new($dyn_send);
+                    ::rustc_data_structures::sync::scope(|s| {
+                        $(
+                            s.spawn(|_| $blocks(*dyn_send));
+                        )*
+                        $fblock(*dyn_send);
+                    })
+                }
             };
-            (impl $fblock:tt [$($blocks:tt,)*] []) => {
-                ::rustc_data_structures::sync::scope(|s| {
+            ([$dyn_send: expr], $fblock:expr, $($blocks:expr),*) => {
+                if is_parallel() {
+                    // Reverse the order of the later blocks since Rayon executes them in reverse order
+                    // when using a single thread. This ensures the execution order matches that
+                    // of a single threaded rustc
+                    parallel!([$dyn_send], impl $fblock [] [$($blocks),*]);
+                } else {
+                    // We catch panics here ensuring that all the blocks execute.
+                    // This makes behavior consistent with the parallel compiler.
+                    let mut ps: Vec<_> = Vec::new();
                     $(
-                        s.spawn(|_| $blocks);
+                        if let Err(p) = ::std::panic::catch_unwind(
+                            ::std::panic::AssertUnwindSafe(|| $blocks($dyn_send))
+                        ) {
+                            ps.push(p);
+                        }
                     )*
-                    $fblock;
-                })
-            };
-            ($fblock:tt, $($blocks:tt),*) => {
-                // Reverse the order of the later blocks since Rayon executes them in reverse order
-                // when using a single thread. This ensures the execution order matches that
-                // of a single threaded rustc
-                parallel!(impl $fblock [] [$($blocks),*]);
+                    ps.into_iter().for_each(|panic| ::std::panic::resume_unwind(panic));
+                }
             };
         }
 
         pub use rayon_core::WorkerLocal;
 
-        pub use rayon::iter::ParallelIterator;
-        use rayon::iter::IntoParallelIterator;
+        use rayon::iter::{FromParallelIterator, IntoParallelIterator, ParallelIterator};
 
-        pub fn par_iter<T: IntoParallelIterator>(t: T) -> T::Iter {
-            t.into_par_iter()
+        pub fn par_for_each_in<I, T: IntoIterator<Item = I> + IntoParallelIterator<Item = I>>(
+            t: T,
+            for_each: impl Fn(I) + DynSync + DynSend
+        ) {
+            if is_parallel() {
+                let for_each = IsSend::new(for_each);
+                let ps: Vec<_> = t.into_par_iter().map(|i| catch_unwind(AssertUnwindSafe(|| for_each(i)))).collect();
```

Comment:
You should not use `collect` here. It is slower than `for_each` and we can only resume a single panic anyway.

Reply:
Yea, I followed your previous [PR](https://github.com/rust-lang/rust/pull/68171) which uses `Lock<Option<_>>` here